hogben/simulate.py

```diff
@@ -10,368 +10,270 @@
-            scale, bkg, dq = probe.intensity, probe.background, probe.dq
-
-            experiment = refl1d_experiment(model.sample, q,
-                                           scale, bkg, dq,
-                                           model.probe.spin_state)
-
-            return experiment.reflectivity()[model.probe.spin_state][1]
-
-        # Otherwise, the sample is not magnetic.
-        else:
-            scale = model.probe.intensity
-            bkg = model.probe.background
-            dq = model.probe.dq
-
-            experiment = refl1d_experiment(model.sample, q, scale, bkg, dq)
-            return experiment.reflectivity()[1]
-
-
-def refl1d_experiment(sample: 'refl1d.model.stack', q_array: np.ndarray,
-                      scale: float, bkg: float, dq: float,
-                      spin_state: Optional[int] = None)\
-                     -> refl1d.experiment.Experiment:
-    """Creates a Refl1D experiment for a given `sample` and `q_array`.
-
-    Args:
-        sample: sample to create an experiment for.
-        q_array: Q points to use in the experiment.
-        scale: experimental scale factor.
-        bkg: level of instrument background noise.
-        dq: instrument resolution.
-        spin_state: spin state to simulate if given a magnetic sample.
-
-    Returns:
-        refl1d.experiment.Experiment: experiment for the given `sample`.
 
-    """
-    # Transform the resolution from refnx to Refl1D format.
-    refl1d_dq = dq / (100*np.sqrt(8*np.log(2)))
-
-    # Calculate the dq array and use it to define a QProbe.
-    dq_array = q_array * refl1d_dq
-    probe = refl1d.probe.QProbe(q_array, dq_array,
-                                intensity=scale, background=bkg)
-    probe.dq = dq
-
-    # Adjust probe calculation for constant dQ/Q resolution.
-    argmin, argmax = np.argmin(q_array), np.argmax(q_array)
-    probe.calc_Qo = np.linspace(q_array[argmin] - 3.5*dq_array[argmin],
-                                q_array[argmax] + 3.5*dq_array[argmax],
-                                21*len(q_array))
-
-    # If the sample is magnetic, create a polarised QProbe.
-    if sample.ismagnetic:
-        probes = [None]*4
-        probes[spin_state] = probe
-        probe = refl1d.probe.PolarizedQProbe(xs=probes, name='')
-        probe.spin_state = spin_state
-
-    return refl1d.experiment.Experiment(probe=probe, sample=sample)
+    non_pol_instr_dict = {'OFFSPEC': 'OFFSPEC_non_polarised_old.dat',
+                          'SURF': 'SURF_non_polarised.dat',
+                          'POLREF': 'POLREF_non_polarised.dat',
+                          'INTER': 'INTER_non_polarised.dat'}
+
+    pol_instr_dict = {'OFFSPEC': 'OFFSPEC_polarised_old.dat',
+                      'POLREF': 'POLREF_polarised.dat'}
+
+    def __init__(self,
+                 sample_model: Union['refnx.reflect.ReflectModel',
+                                     'refl1d.model'],
+                 angle_times: list[tuple],
```

Comment:
It is slightly awkward that now in the Fisher function, in order to calculate the reflectivity, the Simulate class has to be instantiated, where `angle_times` is a mandatory argument. Yet, `angle_times` is not needed in the reflectivity function.

Calling `reflectivity` without actually instantiating the class does not work, as reflectivity is still depended on `self`.

Currently, I work around this by simply using "None" as argument for this:
 ```       
        r = np.concatenate(
            [Simulate(model, None).reflectivity(q) for q, model in
             zip(self.qs, self.structure)])
```

Using dummy values for angle_times seemed less readable than using `None`. As the `None` attribute makes it more clear that this is not used. Could also use an empty list though, as that corresponds with the expected type.